doc/src/devdocs/probes.md

```diff
@@ -90,10 +90,24 @@ the probe handler.
 3. `julia:gc__mark__begin`: Beginning the mark phase
 4. `julia:gc__mark_end(scanned_bytes, perm_scanned)`: Mark phase ended
 5. `julia:gc__sweep_begin(full)`: Starting sweep
-6. `julia:gc__sweep_end()`: Sweep phase finished
+6. `julia:gc__sweep_end`: Sweep phase finished
 7. `julia:gc__end`: GC is finished, other threads continue work
 8. `julia:gc__finalizer`: Initial GC thread has finished running finalizers
 
+### Task runtime probes
+
+1. `julia:rt__run__task(task, ptls)`: Switching to task `task` on thread associated with PTLS `ptls`.
+2. `julia:rt__pause__task(task, ptls)`: Switching from task `task` on thread associated with PTLS `ptls`.
+3. `julia:rt__new__task(parent, child, ptls)`: Task `parent` created task `child` on thread associated with PTLS `ptls`.
+4. `julia:rt__start__process__events(task)`: Task `task` started processing libuv events.
+5. `julia:rt__finish__process__events(task)`: Task `task` finished processing libuv events.
+6. `julia:rt__multiq__insert__success(ptls)`: Thread with PTLS `ptls` successfully inserted a task.
+7. `julia:rt__multiq__insert__full(ptls)`: Thread with PTLS `ptls` failed to insert a task because the multiq was full.
+8. `julia:rt__multiq__deletemin__trylock__success(ptls, heap)`: Thread with PTLS `ptls` successfully acquired the lock for multiq heap `heap`.
+9. `julia:rt__multiq__deletemin__trylock__failed(ptls, heap)`: Thread with PTLS `ptls` failed to acquire a lock. If the number of failed lock attempts significantly exceeds the number of successful locks, it indicates the contention on the (re)scheduling of non-sticky tasks.
+10. `julia:rt__multiq__deletemin__success(ptls, task)`: Thread with PTLS `ptls` successfully acquired task `task` from the multiq.
+11. `julia:rt__multiq__deletemin__emptyish(ptls, task)`: Thread with PTLS `ptls` failed to acquire a task due to the multiq being approximately empty.
```

Comment:
This is pretty comprehensive, but I'm not sure all of them are helpful. Suggestions:
- rename all `multiq` to `taskq` or something general,
- add a `__insert_try` probe (such that `__insert__success` or `__insert__full` are the 'finishes'),
- discard the `trylock` probes,
- rename `deletemin` to `get` or something general,
- add a `__get__try` probe at the `deletemin()` function entry point,
- add probes for all `ptls->sleep_check_state` transitions.

It'd be very helpful to understand where time is spent within [this loop](https://github.com/JuliaLang/julia/blob/8f9b4af796e2374b86dbaef3cb7534f4b032e77d/src/partr.c#L448) but it will take some time to inspect it carefully and make suggestions.

Cc: @tkf

Reply:
I'm OK with dropping `get_next_task` and also the multiq specific ones. These are probably useful more for those who want to optimize the runtime itself rather than analyzing the program written in Julia. But then the former folks can apply the patch themselves. It's probably better to start with a small subset of probes.

> And I think probes at _all_ the sleep state transitions would be helpful rather than just sleep begin and end.

Do you mind sketching how it'd look like? Do you mean other places where `may_sleep` is called?

Reply:
> > And I think probes at _all_ the sleep state transitions would be helpful rather than just sleep begin and end.
> 
> Do you mind sketching how it'd look like? Do you mean other places where `may_sleep` is called?

No, basically all changes to `ptls->sleep_check_state`, e.g. [here](https://github.com/JuliaLang/julia/blob/8f9b4af796e2374b86dbaef3cb7534f4b032e77d/src/partr.c#L363), [here](https://github.com/JuliaLang/julia/blob/8f9b4af796e2374b86dbaef3cb7534f4b032e77d/src/partr.c#L390), [here](https://github.com/JuliaLang/julia/blob/8f9b4af796e2374b86dbaef3cb7534f4b032e77d/src/partr.c#L463), etc.

One goal would be to ensure that the [invariants](https://github.com/JuliaLang/julia/blob/8f9b4af796e2374b86dbaef3cb7534f4b032e77d/src/partr.c#L29-L31) are factual and to guide improvements to the sleep/wakeup logic, which has been difficult. We have a bit of a thundering herd problem right now...

These probes would fall into the same category that you mentioned: useful for developers who want to optimize the runtime itself, so one could argue against adding them to this PR, but I think thread sleep/wakeup is a sufficiently complex and important piece of logic that it should be added now.

Reply:
I see. Yes, I agree it's good to put the probes around these points. If we tweak the code it'd be better to carry the probes.

@jpsamaroo Do you want me to make a patch or do you do it?

Reply:
I dropped the multiq probes on this branch for now (but have them saved). I can introduce the sleep state transitions.

Reply:
Can you all review the sleep state commit and make sure the probes are in the right place, have good names, and pass arguments that make sense?

Reply:
LGTM! There's something I just don't understand how partr.c works https://github.com/JuliaLang/julia/pull/43453#discussion_r780000425 but other than that, this is exactly what I had in mind.